src/PureScript/CST/Utils/ModuleGraph.purs

```diff
@@ -0,0 +1,87 @@
+module PureScript.CST.Utils.ModuleGraph
+  ( moduleGraph
+  , sortModules
+  ) where
+
+import Prelude
+
+import Control.Monad.Free (Free, runFree)
+import Data.Array as Array
+import Data.Foldable (all, foldl)
+import Data.Identity (Identity(..))
+import Data.Map (Map)
+import Data.Map as Map
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Data.Newtype (un)
+import Data.Set (Set)
+import Data.Set as Set
+import Data.Traversable (for)
+import Data.Tuple (Tuple(..))
+import PureScript.CST.Types (ImportDecl(..), Module(..), ModuleHeader(..), ModuleName, Name(..))
+
+type Graph a = Map a (Set a)
+
+moduleGraph :: forall e. Array (Module e) -> Graph ModuleName
+moduleGraph = Map.fromFoldable <<< map go
+  where
+  go (Module { header: ModuleHeader { name: Name { name }, imports } }) =
+    Tuple name (Set.fromFoldable (map getImportName imports))
+
+  getImportName (ImportDecl { module: Name { name } }) = name
+
+sortModules :: forall e. Array (Module e) -> Maybe (Array (Module e))
+sortModules modules = do
+  let
+    moduleNames :: Map ModuleName (Module e)
+    moduleNames =
+      modules
+        # map (\mod@(Module { header: (ModuleHeader { name: Name { name } }) }) -> Tuple name mod)
+        # Map.fromFoldable
+
+    graph = moduleGraph modules
+
+  Array.mapMaybe (flip Map.lookup moduleNames) <$> topoSort graph
+
+topoSort :: forall a. Ord a => Graph a -> Maybe (Array a)
+topoSort graph = do
+  let mbResults = runFree (un Identity) (go (Just { roots: startingModules, sorted: [], usages: importCounts }))
+  map (Array.reverse <<< _.sorted) mbResults
```

Comment:
Can we avoid a reverse if we `cons` instead of `snoc`? Since the number of modules can be quite high, would it be better to use a `List` and then turn it into an Array at the end?

Reply:
Yeah, I'll use a `List` and `Cons` for the internal `topoSort` implementation.